scalableminds/webknossos
# Make Largest Segment ID Optional

Makes the largestSegmentId for segmentation layers optional.
Persists the largestSegmentId for volume annotation layers in the NML

### TODO Backend
 - [x] make largestSegmentId optional in case classes
 - [x] adapt datasource validation
 - [x] create unset value for volumeTracings (cannot use None here because of proto backwards compatibility)
 - [x] volumeTracing largestSegmentId: persist to and read from NML
 - [x] volumeTracing largestSegmentId: make optional

### TODO Frontend
 - [X] when creating a new segment id, check if the volumeTracing’s largestSegmentId is `null`. If so, tell the user about it and provide a way to input a largest segment id
 - [x] Relax the json validation in the dataset edit view
 - [x] Adapt warning message and validation in add-remote-zarr-dataset view

### TODO python client
 - [x] make sure that existing requests don’t break if the segmentation layers do not have a largestSegmentId field https://github.com/scalableminds/webknossos-libs/pull/786

### URL of deployed dev instance (used for testing):
- https://largestsegmentidoptional.webknossos.xyz/auth/login

### Steps to test:
- dataset settings page
  - edit a dataset with a segmentation and set the "largest segment id" field to empty
  - switch to the "advanced" menu and doublecheck that the largest segment id property is either set to null or does not exist
  - add the property there and set it to 10
  - switch back to simple
  - switch back to advanced and remove the property again
  - switch back to simple
  - the field should be empty
  - save
- create an annotation for that dataset (fallback layer should be the segmentation layer without the largest segment id)
  - switch to brush
  - try to brush -> the active segment id should be zero which shouldn't do anything and a toast should warn you about this
  - try to create a new segment id via "c" or the toolbar button --> a toast should appear
  - click the button in the toast and a modal should open which allows to input a largest segment id
  - do so and click ok
  - now, the "c" shortcut should work
- create an annotation without fallback layer --> brushing should just work

### Issues:
- fixes #6039 
- fixes #6390

------
